File: examples/WideningPaper/nested3.R
# This example is a simple nested loop. It has been devised to show
# benefits of the strategy of narrowing with restart. It is the only one that
# can prove that 1 <= i <= 10 inside the inner loop.

nested3 = function() {
  i = 0
  while (TRUE) {
     i = i + 1
     j = 0  
     while (j < 10) 
       j = j + 1   
    if (i > 9) i = 0
  } 
}

File: core/examples/WideningPaper/nested3.c
# Nested 3 in C for use with PAGAI. PAGAI is able to prove that 1 <= i <= 10 in
# the inner loop when using techinque standard with new narrowing.

int main() {
   int i = 0;
   while (1) {
     i = i+1;
     int j = 0;
     while (j < 10)
       j = j+1;
     if (i > 9) i=0;	
   }
}

